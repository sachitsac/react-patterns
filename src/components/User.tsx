import { ReactElement } from "react";
import type { User } from "../types/User";

export const UserComponent = (props: { user: User }): ReactElement => {
  return (
    <div>
      <h1>{props.user.name}</h1>
      <p>{props.user.email}</p>
    </div>
  );
};
