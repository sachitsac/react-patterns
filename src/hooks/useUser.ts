import { useState, useEffect } from "react";
import { getUsers } from "../api/fetchUsers";
import type { User } from "../types/User";

export type UsersResponse = {
  loading: boolean;
  users: User[] | null;
};

export const useUser = () => {
  const [users, setUsers] = useState<User[]>();
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchUsers = async () => {
      const users = await getUsers();
      setUsers(users);
      setLoading(false);
    };
    fetchUsers();
  }, []);
  return {
    loading,
    users,
  };
};
