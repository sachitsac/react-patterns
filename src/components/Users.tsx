import { ReactElement } from "react";
import { useUser } from "../hooks/useUser";
import { UserComponent } from "./User";
import { Loadable } from "./Loadable";

/**
 * This component is used to display the list of users
 * Notice there is only ui logic here and not any business logic
 * We are using a useUser hook to get the list of users from the server
 * This is a presentational component
 *
 * @returns
 */
export const Users = (): ReactElement => {
  const { loading, users } = useUser();

  return (
    <Loadable
      isLoading={loading}
      isEmpty={users?.length === 0}
      renderEmpty={<div>No users available</div>}
    >
      {users?.map((user) => (
        <UserComponent key={user.id} user={user} />
      ))}
    </Loadable>
  );
};
