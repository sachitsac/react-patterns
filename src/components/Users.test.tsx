import { render, screen } from "@testing-library/react";
import { Users } from "./Users";
import * as useUser from "../hooks/useUser";
import { User } from "../types/User";

const users: User[] = [
  {
    id: "user-1",
    name: "John Doe",
    email: "john.doe@example.com",
  },
  {
    id: "user-2",
    name: "Jane Smith",
    email: "jane.smith@example.com",
  },
];

const useUserSpy = vi.spyOn(useUser, "useUser");

const setup = () => {
  render(<Users />);
};

describe("Users", () => {
  beforeEach(() => {
    useUserSpy.mockReturnValue({ loading: false, users });
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it("should render loading message when users are being fetched", () => {
    useUserSpy.mockReturnValue({ loading: true, users: undefined });
    const component = setup();
    expect(screen.getByText("Loading...")).toBeInTheDocument();
    expect(component).toMatchSnapshot();
  });

  it("should render empty message when users are not available", () => {
    useUserSpy.mockReturnValue({ loading: false, users: undefined });
    const component = setup();
    expect(screen.getByText("No users available")).toBeInTheDocument();
    expect(component).toMatchSnapshot();
  });

  it("should render list of users", () => {
    const component = setup();
    users.forEach((user) => {
      expect(screen.getByText(user.name)).toBeInTheDocument();
      expect(screen.getByText(user.email)).toBeInTheDocument();
    });
    expect(component).toMatchSnapshot();
  });
});
