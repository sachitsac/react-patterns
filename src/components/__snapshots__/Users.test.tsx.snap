// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Users > should render empty message when users are not available 1`] = `undefined`;

exports[`Users > should render list of users 1`] = `undefined`;

exports[`Users > should render loading message when users are being fetched 1`] = `undefined`;
